include/swift/SIL/InstructionUtils.h

```diff
@@ -135,6 +135,57 @@ struct LLVM_LIBRARY_VISIBILITY FindClosureResult {
 /// by a reabstraction thunk.
 FindClosureResult findClosureForAppliedArg(SILValue V);
 
+/// Move only data structure that is the result of findLocalApplySite.
+///
+/// NOTE: Generally it is not suggested to have move only types that contain
+/// small vectors. Since our small vectors contain one element or a std::vector
+/// like data structure , this is ok since we will either just copy the single
+/// element when we do the move or perform a move of the vector type.
+struct LLVM_LIBRARY_VISIBILITY FindLocalApplySitesResult {
+  /// Contains the list of local non fully applied partial apply sites that we
+  /// found.
+  SmallVector<ApplySite, 1> partialApplySites;
+
+  /// Contains the list of full apply sites that we found.
+  SmallVector<FullApplySite, 1> fullApplySites;
+
+  /// Set to true if the function-ref_ref escapes into a use that our analysis
+  /// does not understand. Set to false if we found a use that had an actual
+  /// escape. Set to None if we did not find any call sites, but also didn't
+  /// find any "escaping uses" as well.
+  ///
+  /// The none case is so that we can distinguish in between saying that a value
+  /// did escape and saying that we did not find any conservative information.
+  Optional<bool> escapes;
+
+  FindLocalApplySitesResult() = default;
+  FindLocalApplySitesResult(const FindLocalApplySitesResult &) = delete;
+  FindLocalApplySitesResult &
+  operator=(const FindLocalApplySitesResult &) = delete;
+  FindLocalApplySitesResult(FindLocalApplySitesResult &&) = default;
+  FindLocalApplySitesResult &operator=(FindLocalApplySitesResult &&) = default;
+  ~FindLocalApplySitesResult() = default;
+
+  /// Treat this function ref as escaping only if we found an actual user we
+  /// didn't understand. Do not treat it as escaping if we did not find any
+  /// users at all.
+  bool isEscaping() const { return escapes.getValueOr(false); }
+
+  /// We convert to true if we have any "non"-conservative information about the
+  /// FunctionRefInst that was processed.
+  ///
+  /// NOTE: We want to return true if escapes has any value. We want to return
+  /// true if we have any non-conservative information that the user should
+  /// process.
+  operator bool() const {
```

Comment:
The reason that I put this is in was so that I could use an if scope, e.x.:

```
        if (auto result = findLocalApplySites(fri)) {
```

Looks nice, right? But I get your point.

Reply:
I don't get your point. I'm saying you *should* be able to do this:

` if (auto result = findLocalApplySites(fri)) {`

As long as your intention is that the `if` statement executes whenever the `result` is valid.